### PR TITLE
feat(component): on blur of new component name, show dialog of compon…

### LINKED ISSUE #456: Display user advice when creating new component entry
People just create their component, without caring for existing records.

There should be a (Dialogue box) warning which advises the user before creating a new component to really use the search for existing components.

Maybe displaying the warning or not could be a setting in sw360properties.

2. Clarification from Discussion:

Actually having the warning message at "Add Component" could lead to auto-click-away, rather have the proposals when the user enters a new name for the component.

Proposal is to have search once the focus moves away from the name text field and display search results to the user.

If the system found entries, then a dialogue box displays the found entries (best 10).

The user can then either a) ignore or b) jump to the matching component.

### PR DESCRIPTION
**Pull request by imaykay**
…ents with similar name

* added abstract require module that opens a $.confirm dialog displaying a link list requested at a given url
* added this new module to component name field in add AND edit component flow
* added server side that performs the search for similar components
** at the moment this search might be a bit too verbose as it uses the lucene view for components where not only name is searchable
* removed jquery-validation includes as they broke the pages and are already included before each usage of mapEdit
* removed unused local variable

closes #456

### CODE REVIEW

**Comment by mcjaeger** [comment]
actually flow and the integration is good, but the search results are not: for example adding 

* "linux kernel" bring a long list among the 5000 components, including zlib, gnuplot, etc. -> Let us have a a session where

* Adding "tomcat" bring 14 other tomcat components with tomcat in its name (good), but adding "tomcat-apr" returns "No reasonable links found"

more impressions from the testing:

* sort the returned list of hits alphabetically.

* The green rectangle with "here are your links" appears redundant with the heading ("here are similar ones")

* Components can be created without the category filled out. This could be OK, but the red star indicates it is mandatory. I have not paid attention before, but is this changed with this pull request?

* Actually the list itself and the adding of the dialogue is a big improvement, but at testing, it appears like it could be even better to display this as a table (would solve the sorting issue as well) with more information in columns, like component type, home page, vendors, and (first 40 chars of) description

**Comment by mcjaeger** [comment]
(from discussion)
Let's have the link of the component a new tab and skip the table idea (just listing alphabetically the names)

**Comment by imaykay** [comment]
Okay, next iteration:

* search term is split at all characters not in [a-zA-Z_0-9] (= `\W`) and is looked up only in "name" and "vendorNames" of components, with * suffixed each (attention: java uses "_" also as word character, so this one would not result in a split - is this okay?)
* list is sorted alphabetically
* green rectangle now only appears if no similar components are found
* bug with unchecked required fields is fixed (only existed on this branch :-/ )

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/components/ComponentPortlet.java [hunk 1]
```diff
@@ -201,6 +211,89 @@ private void serveAddVendor(ResourceRequest request, ResourceResponse response)
         }
     }
 
+    private void serveCheckComponentName(ResourceRequest request, ResourceResponse response) throws IOException {
+        JSONObject responseJson;
```
[the remark this answers is not in the document]
Comment (imaykay) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/components/ComponentPortlet.java]: now jackson-core is in use, but some dependency to liferay needs to be there for writing to the response :-/

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/components/ComponentPortlet.java [hunk 2]
```diff
@@ -201,6 +211,89 @@ private void serveAddVendor(ResourceRequest request, ResourceResponse response)
...
+        String cName = request.getParameter(PortalConstants.COMPONENT_NAME);
+
+        if (cName != null && !cName.isEmpty()) {
+            List<Component> similarComponents = new ArrayList<>();
+            Map<String, Set<String>> filterMap = new HashMap<>();
+
+            // to find tomcat even on a cName of tomcat-apr, split the cName of special
+            // tokens:
+            // \W = a non word character, so not in [a-zA-Z_0-9]
+            Set<String> splittedCName = Sets.newHashSet(cName.split("\\W"));
```
[the remark this answers is not in the document]
Comment (imaykay) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/components/ComponentPortlet.java]: fixed :-)

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/components/ComponentPortlet.java [hunk 3]
```diff
@@ -201,6 +211,89 @@ private void serveAddVendor(ResourceRequest request, ResourceResponse response)
...
+        responseJson.put("title", "To avoid duplicate components, check these similar ones! Does yours already exist?");
+
+        // adding errors or empty array if none occured
+        JSONArray errorsJsonArray = createJSONArray();
+        errors.stream().forEach(e -> {
+            errorsJsonArray.put(e);
+        });
+        responseJson.put("errors", errorsJsonArray);
+
+        // create link template for component
```
[the remark this answers is not in the document]
Comment (imaykay) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/components/ComponentPortlet.java]: [x] done

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/components/ComponentPortlet.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexbrdn) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/components/ComponentPortlet.java]: That's fine. We just don't want to use Liferay where another library would do.

**Comment by mcjaeger** [comment]
with #544 merged, we could close this, right?

**Comment by imaykay** [comment]
Github marked it automatically as merged, so I would agree.

**Pull request merged**